stock_env.py

```diff
@@ -1,20 +1,125 @@
+            interest_annual (number): annual interest rate
+            borrow_interest_annual (number): annual stock loan fee
+            transaction_penalty (number): transaction fee percentage
+        """
+        days_per_year = 261  # number of trading days per year
+        self.initial_cash = initial_cash
+        self.buy_sell_amt = buy_sell_amt
+        self.exit_threshold = exit_threshold  # minimum total cash value of assets
+        self.inflation = inflation_annual / days_per_year  # daily penalty on all assets
+        self.interest = interest_annual / days_per_year  # daily penalty on borrowed cash
+        self.borrow_interest = borrow_interest_annual / days_per_year  # daily penalty on borrowed stocks
+        self.transaction_penalty = transaction_penalty
+
+        data = get_data()  # pricing data
+        self.pricing_data = data[1] if is_testing else data[0]
+
+    def generate_episode(self, model):
+        """
+        generate an episode of experience based on the model's current policy
+        episode stops when the end of the pricing data is reached, or the agent is broke
+        the environment handles recalculation of cash on hand, cash value of stocks, and total cash value of assets
+
+        :param model: the RL agent, which contains a policy of which actions to take
+
+        :return tuple of lists (states, actions, rewards), where each list is of episode_length.
+                episode_length is not necessarily the same as num_states, because the agent might be broke and
+                terminate the episode early
+        """
+        states = []
+        actions = []
+        rewards = []
+
+        past_num = model.past_num
+        num_stocks = model.num_stocks
+
+        timestep = past_num  # we start on day number <past_num>
+        max_timestep = tf.shape(pricing_data)[1]
+        portfolio_cash = [0] * (num_stocks + 1)  # cash value of each asset
+        portfolio_cash[num_stocks] = self.initial_cash  # cash on hand
+        portfolio_shares = [0] * num_stocks  # shares of each stock owned
+        total_cash_value = self.initial_cash
+
+        # ================ GENERATION ================
+        while total_cash_value > exit_threshold:
+            if timestep > max_timestep:  # we've reached the end of pricing_data
+                break
+
+            sliced_price_history = self.pricing_data[:, timestep -
+                                                     past_num:timestep, :]
+            closing_prices = np.reshape(sliced_price_history[:, past_num, 3],
+                                        (-1, ))
+
+            # recalculate portfolio_cash based on new prices
+            portfolio_cash[:-1] = portfolio_shares * closing_prices
+
+            action = []  # joint action across all stocks
+            transactions = 0  # number of buys and sells
+            state = (sliced_price_history, portfolio_cash)
+            probabilities = model.call([state]).numpy()[0]  # batch_sz=1
```

Comment:
what are u assuming to be the shape of this `probabilities`? asking because I need to determine the output of model.call()

Reply:
model.call should return (batch_sz, num_stocks, 3). here, probabilities = model.call(...)[0] since we just have a single batch, so probabilities is of shape (num_stocks, 3).